### PR TITLE
Jclouds machines: do not persist NodeMetadata or Template

### PR DESCRIPTION
**Pull request by aledsage**
Work in progress:
* requires the same to be done for `JcloudsWinrmMachineLocation`
* requires more testing for backwards compatibility (see the unit test included here; but needs tested in the wild).

Previously JcloudsSshMachineLocation would persist the `NodeMetadata` and the `Template`, which pulled in lots of jclouds stuff. This was very brittle: if the NodeMetadata fields changed, or something like `AzureComputeTemplateOptions` fields changed, then rebind could fail.

This changes `JcloudsSshMachineLocation` to not persist node and template. Instead it persists the nodeId + imageId + privateAddresses + publicAddresses + hostname (i.e. the things in NodeMetadata that are most important).

For backwards compatibility, if on rebind it finds the node in the persisted state then it extracts the fields we need, sets those, and then sets the `node` field to null. Same for the `template` field.

### CODE REVIEW

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsSshMachineLocation.java [hunk 1]
```diff
@@ -146,24 +264,48 @@ public JcloudsLocation getParent() {
...
+                }
+                if (groovyTruth(getPublicAddresses())) {
+                    _privateHostname = getPublicAddresses().iterator().next();
+                } else if (groovyTruth(getHostname())) {
+                    _privateHostname = getHostname();
+                } else {
+                    return null;
+                }
+            }
+        }
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsSshMachineLocation.java]: I think having a single code path might be better here, i.e. have only the fallback? Otherwise could be surprised by different behaviour in some edge cases.

File: locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/RebindJcloudsLocationLiveTest.java [hunk 2]
```diff
@@ -94,6 +101,100 @@ public void testRebindsToJcloudsMachine() throws Exception {
...
+        assertEquals(machine2.getHostname(), "ip-10-144-66-5", errmsg); // TODO would prefer the hostname that works inside and out
+        assertEquals(machine2.getOsDetails().isWindows(), false, errmsg);
+        assertEquals(machine2.getOsDetails().isLinux(), true, errmsg);
+        assertEquals(machine2.getOsDetails().isMac(), false, errmsg);
+        assertEquals(machine2.getOsDetails().getName(), "centos", errmsg);
+        assertEquals(machine2.getOsDetails().getArch(), "x86_64", errmsg);
+        assertEquals(machine2.getOsDetails().getVersion(), "6.5", errmsg);
+        assertEquals(machine2.getOsDetails().is64bit(), true, errmsg);
+
+        // Force it to be persisted again. Expect to pesist without the NodeMetadata and Tempalte.
```
Comment (neykov) [inline locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/RebindJcloudsLocationLiveTest.java]: typo `Tempalte`

**Comment by neykov** [comment]
Really happy to see so much brittle state removed from persistence. Looks good so far.

**Comment by ahgittin** [comment]
+1 ... look forward to having this merged!

**Comment by aledsage** [comment]
@neykov @ahgittin I've done various updates (including getting this working for `JcloudsWinRmMachineLocation`) and fixed some live tests. Can you take another look please?

**Comment by ahgittin** [comment]
looks good
